File: server/api-server/routers/collections.py
"""Collection CRUD endpoints."""
from fastapi import APIRouter, HTTPException, Query

from stdb_client import sql_query, call_reducer, map_collection
from models import (
    CollectionResponse,
    CollectionCreateResponse,
    CollectionUpdateResponse,
    CollectionDeleteResponse,
    PaginatedResponse,
)

router = APIRouter(prefix="/api/v1/collections", tags=["collections"])


@router.get("", response_model=PaginatedResponse)
async def list_collections(
    offset: int = Query(0, ge=0, description="Zero-based offset"),
    limit: int = Query(50, le=100, description="Max results"),
):
    """List all collections."""
    count_rows = await sql_query("SELECT COUNT(*) FROM collection")
    total = count_rows[0][0] if count_rows else 0
    rows = await sql_query("SELECT * FROM collection LIMIT ?i OFFSET ?i", limit, offset)
    return {
        "data": [map_collection(r) for r in rows],
        "total": total,
        "offset": offset,
        "limit": limit,
    }


@router.get("/{collection_id}", response_model=CollectionResponse)
async def get_collection(collection_id: str):
    """Get a single collection by ID."""
    rows = await sql_query("SELECT * FROM collection WHERE id = ?", collection_id)
    if not rows:
        raise HTTPException(404, "Collection not found")
    return map_collection(rows[0])


@router.post("", response_model=CollectionCreateResponse)
async def create_collection(name: str, description: str = "", icon: str = "", color: str = ""):
    """Create a new collection via reducer."""
    result = await call_reducer("create_collection", [name, description, icon, color])
    return result or {"status": "created"}


@router.put("/{collection_id}", response_model=CollectionUpdateResponse)
async def update_collection(collection_id: str, name: str | None = None, description: str | None = None):
    """Update a collection via reducer."""
    current_name = name or ""
    current_desc = description or ""
    await call_reducer("update_collection", [collection_id, current_name, current_desc, "", ""])
    return {"status": "updated"}


@router.delete("/{collection_id}", response_model=CollectionDeleteResponse)
async def delete_collection(collection_id: str):
    """Delete a collection via reducer."""
    await call_reducer("delete_collection", [collection_id])
    return {"status": "deleted"}

File: server/api-server/src/routes/collections.js
import { Router } from 'express';
import { sqlQuery, callReducer, genId, mapCollection } from '../stdb.js';

const router = Router();

// GET /api/v1/collections — List all collections
router.get('/', async (req, res, next) => {
  try {
    const rows = await sqlQuery('SELECT * FROM collection ORDER BY sort_order ASC');
    res.json({ data: rows.map(mapCollection) });
  } catch (err) { next(err); }
});

// GET /api/v1/collections/:id — Get a single collection
router.get('/:id', async (req, res, next) => {
  try {
    const rows = await sqlQuery(`SELECT * FROM collection WHERE id = '${req.params.id}'`);
    if (rows.length === 0) return res.status(404).json({ error: 'Collection not found' });
    res.json({ data: mapCollection(rows[0]) });
  } catch (err) { next(err); }
});

// POST /api/v1/collections — Create a new collection
router.post('/', async (req, res, next) => {
  try {
    const { name, description, parent_id, icon, color } = req.body;
    if (!name) return res.status(400).json({ error: 'name is required' });
    const id = genId('col');
    await callReducer('create_collection', [
      id, name, description || '', parent_id || '', icon || '', color || '', req.apiUser.id,
    ]);
    const rows = await sqlQuery(`SELECT * FROM collection WHERE id = '${id}'`);
    res.status(201).json({ data: rows.length ? mapCollection(rows[0]) : { id } });
  } catch (err) { next(err); }
});

// PUT /api/v1/collections/:id — Update a collection
router.put('/:id', async (req, res, next) => {
  try {
    const { name, description, icon, color } = req.body;
    const existing = await sqlQuery(`SELECT * FROM collection WHERE id = '${req.params.id}'`);
    if (existing.length === 0) return res.status(404).json({ error: 'Collection not found' });

    await callReducer('update_collection', [
      req.params.id,
      name || String(existing[0][1] ?? ''),
      description || String(existing[0][3] ?? ''),
      icon ?? String(existing[0][5] ?? ''),
      color ?? String(existing[0][6] ?? ''),
    ]);
    const rows = await sqlQuery(`SELECT * FROM collection WHERE id = '${req.params.id}'`);
    res.json({ data: mapCollection(rows[0]) });
  } catch (err) { next(err); }
});

// DELETE /api/v1/collections/:id — Delete a collection
router.delete('/:id', async (req, res, next) => {
  try {
    await callReducer('delete_collection', [req.params.id]);
    res.json({ success: true });
  } catch (err) { next(err); }
});

export default router;
